### DISCUSSION TITLE
Azure devops automerge pulling branch instead of PR status

### DISCUSSION THREAD

**Discussion by rkrabek** [renovatebot/renovate#17558] [Request Help]
Using latest Renovate via npx
Hosted on dev.azure.com

It appears from the debug logs that renovate is pulling the branch status rather than the PR status:
`2022-08-31T20:27:54.2486423Z DEBUG: getBranchStatus(renovate/tether-1.4.x) (repository=TestRepo/RenovateTest, branch=renovate/tether-1.4.x)
2022-08-31T20:27:55.0399072Z DEBUG: branch status check result (repository=TestRepo/RenovateTest, branch=renovate/tether-1.4.x)
2022-08-31T20:27:55.0399900Z        "statuses": []
2022-08-31T20:27:55.0401046Z DEBUG: empty branch status check result = returning "pending" (repository=TestRepo/RenovateTest, branch=renovate/tether-1.4.x)
2022-08-31T20:27:55.0402199Z DEBUG: Branch status yellow (repository=TestRepo/RenovateTest, branch=renovate/tether-1.4.x)
2022-08-31T20:27:55.0403379Z DEBUG: PR is not ready for merge (branch status is yellow) (repository=TestRepo/RenovateTest, branch=renovate/tether-1.4.x)`

![image](https://user-images.githubusercontent.com/9980031/187777894-ec528098-01fd-4720-8f7c-4ef8fb75634d.png)

From [this](https://github.com/renovatebot/renovate/issues/7392) PR it would appear Renovate is using this api: https://docs.microsoft.com/en-us/rest/api/azure/devops/git/statuses?view=azure-devops-rest-4.1
rather than the PR status api: https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull-request-statuses/list?view=azure-devops-rest-6.0&tabs=HTTP
so the debug message is a little misleading when it says the PR is not ready to merge because the branch status is yellow. 

For the most part it's a non-issue if you turn on platform automerge so I'm not sure it's worth the effort to implement the PR status api specifically for the azure platform but it may help updating the debug logs for anyone looking at the logs and trying to figure out why their automerge isn't working.

**Comment by PhilipAbed** [comment]
In all platforms we check if the branch checks are done.. things like build/tests and other checks.
PR is created from a branch so the branch status is what we are looking for to let do automerge.

We do not support optional checks yet i believe i saw a feature called "auto merge configuration" but it is not fully implemented yet
see https://github.com/renovatebot/renovate/issues/359


Are you sure it is better to use the PR Statuses api instead?
can you call the 2 APIs and show me the results?

**Reply by rkrabek** [reply]
So I believe the disconnect here is that at least in azure the PR statuses are separate from the branch status. "Branch status" is kind of a misnomer from my understanding because the [API ](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/statuses?view=azure-devops-rest-4.1) used by renovate is actually pulling the status of a particular commit rather than the status of a PR that has the source branch already merged with the target branch.

For a renovate PR that has statuses looking like this:
![image](https://user-images.githubusercontent.com/9980031/188242083-9dcbd10d-5842-4601-8706-562b52deca38.png)
Using the commit status [API](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/statuses/list?view=azure-devops-rest-6.0&tabs=HTTP) I get:
`{"count":0,"value":[]}`
Using the PR status [API](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull-request-statuses/list?view=azure-devops-rest-6.0&tabs=HTTP) I get 400 lines of output but I'll put a small sample below:
`{
    "value": [
      {
        "iterationId": 1,
        "id": 1,
        "description": "Coverage status check queued for RenovateTest - PR Validation",
        "context": {
          "name": "codecoverage",
          "genre": "renovatetest - pr validation"
        },
        "creationDate": "2022-08-31T19:04:34.2927741Z",
        "updatedDate": "2022-08-31T19:04:34.2927741Z",
        "createdBy": {
          "displayName": "Azure Pipelines Test Service",
          "url": "a url here",
          "_links": {
            "avatar": {
              "href": "other urls here"
            }
          },
          "id": "id here",
          "uniqueName": "",
          "imageUrl": "other urls here",
          "descriptor": "blah"
        }
      },
      {
        "iterationId": 1,
        "id": 2,
        "state": "pending",
        "description": "Nitpicker",
        "context": {
          "name": "comment-status",
          "genre": "pr-nitpicker"
        },
        "creationDate": "2022-08-31T19:04:57.8866704Z",
        "updatedDate": "2022-08-31T19:04:57.8866704Z",
        "createdBy": {
          "displayName": "MerlinBot",
          "url": "url here",
          "_links": {
            "avatar": {
              "href": "other url here"
            }
          },
          "id": "id here",
          "uniqueName": "",
          "imageUrl": "image url here",
          "descriptor": "descriptor here"
        },
        "targetUrl": "targeturl of the  nitpicker"
      },`

I am able to get renovatebot to register something from the branch status API if I run a particular pipeline against the branch itself like below
![image](https://user-images.githubusercontent.com/9980031/188241804-bde4ddcc-8d73-418f-a484-f19d67f4486a.png)
But this kind of defeats the purpose of automerge checking the PR status because it will register that this the only status that the commit has and that it has passed while ignoring the other status checks that happen at PR time configured by branch policies.

Another way to see this is to go to the commits tab, select the branch that renovate made a PR from and look at the status column:
![image](https://user-images.githubusercontent.com/9980031/188242570-721256d2-0196-4c92-95f4-018ce741d436.png)
In this case there is a currently open PR from renovate for babel/monorepo but under the status column only the previously merged commits have statuses associated with them (from the PR build prior to merging I'm assuming) whereas the newest commit doesn't have any statuses associated with it despite the PR status checks already having run

**Comment by rarkins** [comment]
This would affect regular Renovate automerge. Happy if you would like to raise a feature request to check Azure PR status
